btcstaking-tracker/stakingeventwatcher/stakingeventwatcher.go

```diff
@@ -0,0 +1,551 @@
+		if txIn.PreviousOutPoint.Hash == stakingTxHash && txIn.PreviousOutPoint.Index == td.StakingOutputIdx {
+			return i, nil
+		}
+	}
+
+	return -1, fmt.Errorf("transaction does not point to staking output. expected hash:%s, expected outoputidx: %d", stakingTxHash, td.StakingOutputIdx)
+}
+
+// tryParseStakerSignatureFromSpentTx tries to parse staker signature from unbonding tx.
+// If provided tx is not unbonding tx it returns error.
+func tryParseStakerSignatureFromSpentTx(tx *wire.MsgTx, td *TrackedDelegation) (*schnorr.Signature, error) {
+	if len(tx.TxOut) != 1 {
+		return nil, fmt.Errorf("unbonding tx must have exactly one output. Priovided tx has %d outputs", len(tx.TxOut))
+	}
+
+	if tx.TxOut[0].Value != td.UnbondingOutput.Value || !bytes.Equal(tx.TxOut[0].PkScript, td.UnbondingOutput.PkScript) {
+		return nil, fmt.Errorf("unbonding tx must have ouput which matches unbonding output of retrieved from Babylon")
+	}
+
+	stakingTxInputIdx, err := getStakingTxInputIdx(tx, td)
+
+	if err != nil {
+		return nil, fmt.Errorf("unbonding tx does not spend staking output: %v", err)
+	}
+
+	stakingTxInput := tx.TxIn[stakingTxInputIdx]
+	witnessLen := len(stakingTxInput.Witness)
+	// minimal witness size for staking tx input is 4:
+	// covenant_signature, staker_signature, script, control_block
+	// If that is not the case, something weird is going on and we should investigate.
+	if witnessLen < 4 {
+		panic(fmt.Errorf("staking tx input witness has less than 4 elements for unbonding tx %s", tx.TxHash()))
+	}
+
+	// staker signature is 3rd element from the end
+	stakerSignature := stakingTxInput.Witness[witnessLen-3]
+
+	return schnorr.ParseSignature(stakerSignature)
+}
+
+// waitForDelegationToStopBeingActive polls babylon until delegation is no longer active.
+func (sew *StakingEventWatcher) waitForDelegationToStopBeingActive(
+	ctx context.Context,
+	stakingTxHash chainhash.Hash,
+) {
+	_ = retry.Do(func() error {
+		active, err := sew.babylonNodeAdapter.IsDelegationActive(stakingTxHash)
+
+		if err != nil {
+			return fmt.Errorf("error checking if delegation is active: %v", err)
+		}
+
+		if !active {
+			return nil
+		}
+
+		return fmt.Errorf("delegation for staking tx %s is still active", stakingTxHash)
+	},
+		retry.Context(ctx),
+		retryForever,
+		fixedDelyTypeWithJitter,
+		retry.MaxDelay(sew.cfg.CheckDelegationActiveInterval),
+		retry.MaxJitter(sew.cfg.RetryJitter),
+		retry.OnRetry(func(n uint, err error) {
+			sew.logger.Debugf("retrying checking if delegation is active for staking tx %s. Attempt: %d. Err: %v", stakingTxHash, n, err)
+		}),
+	)
+}
+
+func (sew *StakingEventWatcher) reportUnbondingToBabylon(
+	ctx context.Context,
+	stakingTxHash chainhash.Hash,
+	unbondingSignature *schnorr.Signature,
+) {
+	_ = retry.Do(func() error {
+		active, err := sew.babylonNodeAdapter.IsDelegationActive(stakingTxHash)
+
+		if err != nil {
+			return fmt.Errorf("error checking if delegation is active: %v", err)
+		}
+
+		if !active {
```

Comment:
we should report unbonding transaction to Babylon even if it not active i.e babylon will accept unbonding transactions report if delegation is:
- either active
- or verified
and we should stop reporting if it is unbonded